### PR TITLE
swhkd: init at unstable-2023-10-16

### PR DESCRIPTION
**Pull request by ivandimitrov8080**
## Description of changes

Adding new package SWHKD

## Things done

Defined the derivation

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sw/swhkd/package.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, pkgs
+, stdenv
+, fetchFromGitHub
+, rustPlatform
+}:
+let
+  name = "swhkd";
+  version = "30f25b5bf99df5f16d91b51a7bd397c1de075085";
+  buildInputs = with pkgs; [ systemd ];
```
Comment (KAction) [inline pkgs/by-name/sw/swhkd/package.nix:10]: From what I glanced at upstream page, `systemd` should not be hard dependency.

Consider moving systemd-related stuff into separate output and adding `withSystemd` flag for the benefit of platforms that do not have systemd (e.g static platform).

File: pkgs/by-name/sw/swhkd/package.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  inherit name buildInputs nativeBuildInputs;
+  src = fetchFromGitHub {
+    owner = "waycrate";
+    repo = name;
+    rev = version;
+    hash = "sha256-K05v2xw7G/ZJsf62g696TOSZHMAOWf24sT9JcvvBNx0=";
+  };
+  cargoHash = "sha256-E5AE18CfeX1HI/FbGDFoUDsPyG/CpJrD+8Ky7c+EQUw=";
+  meta = {
+    description = "Sxhkd clone for Wayland (works on TTY and X11 too)";
```
Comment (KAction) [inline pkgs/by-name/sw/swhkd/package.nix:23]: Not everybody knows what is `Sxhkd`. Consider replacing with `A hotkey daemon for Wayland, X11 and TTY`.

File: pkgs/by-name/sw/swhkd/package.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    homepage = "https://github.com/waycrate/swhkd";
```
Comment (KAction) [inline pkgs/by-name/sw/swhkd/package.nix:24]: Consider also linking changelog https://github.com/waycrate/swhkd/blob/main/CHANGELOG.md

File: pkgs/by-name/sw/swhkd/package.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+, rustPlatform
+}:
+let
+  name = "swhkd";
+  version = "30f25b5bf99df5f16d91b51a7bd397c1de075085";
+  buildInputs = with pkgs; [ systemd ];
+  nativeBuildInputs = with pkgs; [ pkg-config ];
+in
+rustPlatform.buildRustPackage rec {
+  inherit name buildInputs nativeBuildInputs;
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:14]:
```suggestion

rustPlatform.buildRustPackage rec {
  pname = "swhkd";
  version = "unstable-2023-10-16";

```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  src = fetchFromGitHub {
+    owner = "waycrate";
+    repo = name;
+    rev = version;
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:18]:
```suggestion
    rev = "30f25b5bf99df5f16d91b51a7bd397c1de075085";;
```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  name = "swhkd";
+  version = "30f25b5bf99df5f16d91b51a7bd397c1de075085";
+  buildInputs = with pkgs; [ systemd ];
+  nativeBuildInputs = with pkgs; [ pkg-config ];
+in
+rustPlatform.buildRustPackage rec {
+  inherit name buildInputs nativeBuildInputs;
+  src = fetchFromGitHub {
+    owner = "waycrate";
+    repo = name;
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:17]:
```suggestion
    repo = "swhkd";
```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    rev = version;
+    hash = "sha256-K05v2xw7G/ZJsf62g696TOSZHMAOWf24sT9JcvvBNx0=";
+  };
+  cargoHash = "sha256-E5AE18CfeX1HI/FbGDFoUDsPyG/CpJrD+8Ky7c+EQUw=";
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:21]:
```suggestion

  cargoHash = "sha256-E5AE18CfeX1HI/FbGDFoUDsPyG/CpJrD+8Ky7c+EQUw=";

  nativeBuildInputs = [ pkg-config ];

  buildInputs = [ systemd ];
```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 8]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, pkgs
+, stdenv
+, fetchFromGitHub
+, rustPlatform
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:5]:
```suggestion
, fetchFromGitHub
, rustPlatform
, pkg-config
, systemd
```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    owner = "waycrate";
+    repo = name;
+    rev = version;
+    hash = "sha256-K05v2xw7G/ZJsf62g696TOSZHMAOWf24sT9JcvvBNx0=";
+  };
+  cargoHash = "sha256-E5AE18CfeX1HI/FbGDFoUDsPyG/CpJrD+8Ky7c+EQUw=";
+  meta = {
+    description = "Sxhkd clone for Wayland (works on TTY and X11 too)";
+    homepage = "https://github.com/waycrate/swhkd";
+    license = lib.licenses.bsd2;
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:25]: Please add yourself as a maintainer by adding an alphabetically sorted entry to `maintainers/maintainer-list.nix` and commit this change into a separate commit with the message ``maintainers: add NAME`` into this PR.

File: pkgs/by-name/sw/swhkd/package.nix [hunk 10]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, pkgs
+, stdenv
+, fetchFromGitHub
+, rustPlatform
+}:
+let
+  name = "swhkd";
+  version = "30f25b5bf99df5f16d91b51a7bd397c1de075085";
```
Comment (KAction) [inline pkgs/by-name/sw/swhkd/package.nix:9]: Convention for version is `1.2.1+unstable=YYYY-MM-DD`, with date corresponding to the commit date. Also consider asking upstream for proper tag.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
The commit title should be `swhkd: init at unstable-2023-10-16`

**Review by KAction — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sw/swhkd/package.nix (hunk 1 above)
Comment (ivandimitrov8080) [reply pkgs/by-name/sw/swhkd/package.nix:10]: Hey,

[This answer](https://discourse.nixos.org/t/solved-libudev-replaced-by-udev/18951/3?u=ivandimitrov8080) showed me how to fix a build failure. If I don't specify systemd it doesn't build.

**Comment by donovanglover** [comment]
Related issue: https://github.com/NixOS/nixpkgs/issues/180942
Prior work: https://github.com/NixOS/nixpkgs/pull/255374

File: pkgs/by-name/sw/swhkd/package.nix (hunk 1 above)
Comment (KAction) [reply pkgs/by-name/sw/swhkd/package.nix]: Ah, I see. Can you then please add `meta.broken = !(lib.meta.availableOn stdenv.hostPlatform.systemd)`. Will make error cleaner for somebody who will try to build on unsupported system.

**Review by khaneliman — CHANGES_REQUESTED** [review]
Reason why we are ignoring the original PR for the new package? I see that one is pointing at a 1.2.2 branch and this is latest commit on main. Should we have a stable and unstable or are you just doing unstable? 

Maintainer add commit should be before the new package commit and squash the first 2 commits into one 'swhkd: init at unstable-2023-10-16' after it...

File: pkgs/by-name/sw/swhkd/package.nix [hunk 11]
```diff
@@ -0,0 +1,29 @@
+{ lib
+, pkgs
+, fetchFromGitHub
+, rustPlatform
+, pkg-config
+}:
+let
+in
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:8]:
```suggestion

```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  src = fetchFromGitHub {
+    owner = "waycrate";
+    repo = "swhkd";
+    rev = "30f25b5bf99df5f16d91b51a7bd397c1de075085";
+    hash = "sha256-K05v2xw7G/ZJsf62g696TOSZHMAOWf24sT9JcvvBNx0=";
+  };
+  cargoHash = "sha256-mHLrjNVgB5Fxo1Iv4VuHd7ZWQBIiCQK7A+BOfrYdR54=";
+  buildInputs = with pkgs; [ systemd ];
+  nativeBuildInputs = with pkgs; [ pkg-config ];
+  meta = {
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:21]:
```suggestion
  version = "unstable-2023-10-16";

  src = fetchFromGitHub {
    owner = "waycrate";
    repo = "swhkd";
    rev = "30f25b5bf99df5f16d91b51a7bd397c1de075085";
    hash = "sha256-K05v2xw7G/ZJsf62g696TOSZHMAOWf24sT9JcvvBNx0=";
  };

  cargoHash = "sha256-mHLrjNVgB5Fxo1Iv4VuHd7ZWQBIiCQK7A+BOfrYdR54=";

  nativeBuildInputs = [ pkg-config ];

  buildInputs = [ systemd ];
  
  meta = {
```

File: pkgs/by-name/sw/swhkd/package.nix [hunk 13]
```diff
@@ -0,0 +1,29 @@
+{ lib
+, pkgs
+, fetchFromGitHub
+, rustPlatform
+, pkg-config
```
Comment (SuperSandro2000) [inline pkgs/by-name/sw/swhkd/package.nix:5]:
```suggestion
, fetchFromGitHub
, rustPlatform
, pkg-config
, systemd
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Using merge commits in PRs is not that nice as that makes resolving merge conflicts if they happen to be on those commits, near impossible.

Closing in favor of #299648

**Pull request closed without merging**